core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -209,18 +209,21 @@ abstract class AbstractFetcherThread(name: String,
     val partitionsWithError = mutable.HashSet.empty[TopicPartition]
 
     for (tp <- partitions) {
-      try {
-        val highWatermark = partitionStates.stateValue(tp).fetchOffset
-        val truncationState = OffsetTruncationState(highWatermark, truncationCompleted = true)
-
-        info(s"Truncating partition $tp to local high watermark $highWatermark")
-        truncate(tp, truncationState)
-
-        fetchOffsets.put(tp, truncationState)
-      } catch {
-        case e: KafkaStorageException =>
-          info(s"Failed to truncate $tp", e)
-          partitionsWithError += tp
+      val partitionState = partitionStates.stateValue(tp)
+      if (partitionState != null) {
```

Comment:
nit: Looking at the usage in the file, I think we would prefer:
```
Option(partitionStates.stateValue(topicPartition)).foreach { currentFetchState =
...
```
instead of the null check